### ISSUE TITLE
getting the IR for library functions

### ISSUE THREAD

**Issue by Felkin1** [numba/numba#4356]
I've a project where I am trying to get the 'full' LLVM IR of some python code calling different libraries like sklearn, numpy and tensorflow. Is my understanding correct that I have to fork and @jit ALL of their functions for them to show up in the dump? Or am I missing some option to give the full view.

**Comment by sklam** [comment]
You don't have to fork the other projects to provide new jit-able definitions.  Extensions can use the `@overload` decorator.  For details see [the guide for `@overload`](http://numba.pydata.org/numba-doc/latest/extending/overloading-guide.html).  

Also, Numba support only a subset of Python. It is very likely that the code from sklearn and tensorflow will need to be rewritten to work in Numba.

**Comment by Felkin1** [comment]
By work, do you mean to be optimizable or to generate the IR too? My only goal is to get the llvm ir / asm dumps with maximum function call depth.

**Comment by esc** [comment]
@Felkin1 thanks for asking this question and for helping to improve Numba.

It isn't quite clear to me what you are looking for, but I will attempt to answer your question anyway. Currently Numba supports only a limited subset of Python and Numpy out of the box. There is some initial support for Scipy in the  Scumba project: https://github.com/numba/scumba. In order to get what you want you will probably have to `@jit` or `@overload` all the functions you call from those libraries you mentioned.

**Comment by Felkin1** [comment]
@esc Essentially, I am trying to run some performance analysis on some python projects. For that analysis I want to look at what specific operations are 'hot' in the code. I want to see exactly what happens at these hot places and so cprofiler is not perfect. 

I was thinking to use Numba to just scan all the code and get the LLVM IR representation. I want to see what happens not only in the udf, but also the the specific numpy functions.

So it's all on only the analysis part to get the representation, I don't care about the actual optimization Numba could do on the code so I was wondering if the 'lack of support' includes what Numba can accurately build into IR.

**Comment by seibert** [comment]
I don't think Numba will do what you are expecting.  When Numba compiles a function, we substitute in new implementations for functions that we have coded in a way that can be translated to LLVM IR.  The NumPy implementation is not used in any way.  (Exception is that we do delegate linear algebra calls to the BLAS library)  Any analysis you do of the code Numba generates will reflect what is happening in the original application.

That said, you might find this experiment we did useful:

https://github.com/numba/stacktrace

This package allows you to build a sampling profiler that captures both the Python and C stack:

https://stacktrace.readthedocs.io/en/latest/tools.html
